### ISSUE TITLE
Add callbacks to rust

### ISSUE THREAD

**Issue by hce** [gluon-lang/gluon#744]
Another question: I'd like to optimize performance of the script by moving parts of it to rust, especially since BTreeMap serialization to gluon is pretty slow. So what I'd like to do from gluon is this:

rustlib.matchon("^MY_STRING$", \matched string -> ...)

So for every call to matchon, a Regex should be compiled on rust side and whenever it matches, the function passed as the 2nd argument should be called with the matched string. The first idea is to pass a mutable vector to gluon, but that doesn't work. What are your suggestions to achieve the above functionality?

**Comment by Marwes** [comment]
> `\matched string -> ...)`

This is a gluon closure, right? Which would perhaps take an array of strings? Or do you mean that closure would take a mutable vector?

**Comment by hce** [comment]
Yes, a gluon closure that gets called for every match. And yes, either as a string or an array of strings (Vec<String> on the rust side). I don't see why the vector should be mutable, though.

**Comment by Marwes** [comment]
You can use `https://docs.rs/gluon_vm/0.11.2/gluon_vm/api/function/type.OwnedFunction.html` (or `FunctionRef` maybe in this case for slightly smaller overhead) as the argument on the rust side of `matchon`. Then just call `call` to pass whatever argument you want back to the gluon closure.

**Comment by hce** [comment]
Yeah, ok, but I need to add that FunctionRef to a vector on the rust side by calling push. And I don't see how to do that, given that you can not pass closures to gluon.

let mut callbacks: Vec<(Regex, FunctionRef)> = Vec::new();
fn gluon_match_for(String regex, FunctionRef<fn (String) -> []>) {
   // how to reference callbacks here?
}

**Comment by Marwes** [comment]
Ah, I would add a context parameter to `gluon_match_for`

```
#[derive(VmType, Userdata)]
#[gluon(vm_type = "my_module.Context")] // Needs a call to `vm.register_type::<Context>("my_module.Context", [])?`
struct Context(Mutex<Vec<OwnedFunction<fn (String) -> []>>>);
fn gluon_match_for(context: &Context, regex: String, FunctionRef<fn (String) -> []>)
```

If you don't want to pass that around on the gluon side you may then curry the call with that parameter as in https://github.com/gluon-lang/gluon/issues/742#issuecomment-497983040

**Comment by hce** [comment]
I'm getting this error when trying to declare the Context type:

    26 | #[derive(VmType, Userdata)]
       |          ^^^^^^  -------- first implementation here
       |          |
       |          conflicting implementation for `ct::Context`

**Comment by Marwes** [comment]
Right, that fix isn't released yet. Deriving `UserData` automatically implements `VmType` in the version on crates.io. I think you can just remove the `VmType` derive for now

**Comment by Marwes** [comment]
Or use the git version

**Comment by hce** [comment]
I tried to omit VmType, got this error:
^^^^^^^^^^^^^^ the trait `gluon_vm::api::Getable<'_, '_>` is not implemented for `my::Context`

I switched to the git version (commit 45427d5d1b7cde9ca4160a2b2d672bf4a6003a68), added VmType, got the same error...

**Comment by Marwes** [comment]
Derving just VmType and Userdata should be enough on got master https://github.com/gluon-lang/gluon/blob/master/repl/src/repl.rs#L193

**Comment by hce** [comment]
I'm sorry, my bad. Of course, I should have used a reference to the context in the functions exported to gluon. It now works fine with the master branch. (I had to implement the Debug trait also)
